errors.md

```diff
@@ -0,0 +1,448 @@
+# Error Handing in Go for Java Authors
+
+This document compares error handling methodologies in Go and Java.  It holds the view
+that there are more commonalities than not between the two languages in terms
+of capabilities, approaches, and level of expedience.
+
+The analysis is concerned solely with the matters of having a programmer programmatically deal
+with error conditions, not with operators debugging problems at runtime.  This runtime problem
+diagnosis use case is perfectly valid but beyond the scope of this document.
+
+## Cheap and Easy
+
+On the cheapest end of the spectrum, one can use stringly-typed errors in both
+languages:
+
+```java
+void compute(Argument arg) {
+  if (!arg.isValid()) {
+    throw new IllegalArgumentException("arg is invalid.");
+  }
+  // Rest omitted.
+}
+
+```
+
+```go
+func Compute(arg Argument) error {
+  if !arg.valid() {
+    return fmt.Errorf("arg is invalid.")
+  }
+  // Rest omitted.
+}
+```
+
+Both errors contain some information for a human to debug the problem from logs and
+middleware handlers. But all their callers can do (at runtime) is merely detect that
+there was an error, potentially the problem class.  In neither case can the program
+react to the error beyond simple happy and sad paths.
+
+For some cases the boolean error case may be acceptable.
+
+## Discriminating Error Reasons
+
+In contrast to the _Cheap and Easy_ approach, one may require the call graph to take
+decisive and intelligent action based on a specific error condition and provide general handling for everything else.
+
+```java
+void debitAccount(Integer accountId, Integer amount) throws AuthorizationException {
+  // Only agent may debit the principal's account.
+  if (!isDebitAuthorized(this.agentId, accountId)) {
+    throw new AuthorizationException(this.agentId, accountId, "debit");
+  }
+  // Rest omitted.
+}
+
+class AuthorizationException extends Exception {
+  private Integer agentId;
+  private Integer accountId;
+  private String operation;
+  
+  AuthorizationException(Integer agentId, Integer accountId, String operation) {
+    this.agentId = agentId;
+	this.accountId = accountId;
+	this.operation = operation;
+  }
+  
+  public String toString() {
+    return "AuthorizationException(" + agentId + " is not authorized to " + opertation + " account" + accountId;
+  }
+}
+```
+
+```go
+func debitAccount(agentId, accountId, amount int) error {
+  if !isDebitAuthorized(agentId, accountId) {
+    return &AuthorizationError{agentId: agentId, accountId: accountId, operation: "debit"}
+  }
+  // Rest omitted.
+}
+
+type AuthorizationError struct {
+  agentId, accountId int
```

Comment:
Addendum: I see you addressed this in more detail below.

Reply:
Yep.  I want treat this area as a — how do I a put it — as a climb up a ladder of correctness and effort.  Considering how many developers (of any language stripe) are of the `if err != nil` mindset, we need to spell out each step up the ladder incrementally.

Reply:
Sure, and that makes perfect sense. But then: Why bother having fields here at all? At this stage, there's nothing very interesting going on with the internals of the type, and a reasonable person might be asking "why bother storing these as fields rather than just constructing the error string once and be done? I can't do anything with those data."